File: matlab/ofec_single/ebn0_to_sigma.m
function sigma = ebn0_to_sigma(ebn0_dB, bits_per_symbol, code_rate)
%EBN0_TO_SIGMA Convert Eb/N0 to Gaussian std dev (per dimension).
if bits_per_symbol <= 0
    error('ebn0_to_sigma:InvalidM', 'bits_per_symbol must be > 0.');
end
if code_rate <= 0 || code_rate > 1
    error('ebn0_to_sigma:InvalidRate', 'code_rate must lie in (0,1].');
end
ebn0_lin = 10^(ebn0_dB / 10);
esn0_lin = ebn0_lin * bits_per_symbol * code_rate;
N0 = 1 / esn0_lin;
sigma = sqrt(N0 / 2);
end

File: new_float_only/src/channel/ebn0_to_sigma.cpp
#include "new_float_only/channel/awgn.hpp"
#include <cmath>
#include <stdexcept>

namespace channel {

float ebn0_to_sigma(float ebn0_dB, unsigned bits_per_symbol, float code_rate)
{
    if (bits_per_symbol == 0) throw std::invalid_argument("bits_per_symbol must be > 0");
    if (code_rate <= 0.f || code_rate > 1.f) throw std::invalid_argument("code_rate must be in (0,1]");

    const float ebn0_lin = std::pow(10.0f, ebn0_dB / 10.0f);
    const float esn0_lin = ebn0_lin * (bits_per_symbol);
    const float N0       = 1.0f / esn0_lin;      // 因为 Es=1
    const float sigma    = std::sqrt(N0 * 0.5f); // 每一维方差 N0/2
    return sigma;
}

} // namespace new_float_only
